drivers/sensor/tmag5273/tmag5273.c

```diff
@@ -0,0 +1,1118 @@
+		       TMAG5273_CONVERSION_COMPLETE) {
+			retval = i2c_reg_read_byte_dt(
+				&drv_cfg->i2c, TMAG5273_REG_CONV_STATUS | conv_bit, &i2c_buffer[0]);
+
+			if (retval < 0) {
+				LOG_ERR("error reading conversion state %d", retval);
+				return retval;
+			}
+
+			conv_bit = 0;
+
+			k_usleep(drv_data->conversion_time_us);
+		}
+	}
+
+	/* read data */
+	uint8_t start_address = 0;
+	uint8_t end_address = 0;
+
+	switch ((uint16_t)chan) {
+	case SENSOR_CHAN_MAGN_X:
+		if (!drv_cfg->x_axis) {
+			LOG_ERR("x-axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_X_MSB_RESULT;
+		end_address = TMAG5273_REG_X_LSB_RESULT;
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		if (!drv_cfg->y_axis) {
+			LOG_ERR("y-axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_Y_MSB_RESULT;
+		end_address = TMAG5273_REG_Y_LSB_RESULT;
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		if (!drv_cfg->z_axis) {
+			LOG_ERR("x-axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_Z_MSB_RESULT;
+		end_address = TMAG5273_REG_Z_LSB_RESULT;
+		break;
+	case SENSOR_CHAN_MAGN_XYZ:
+		if (!drv_cfg->x_axis && !drv_cfg->y_axis && !drv_cfg->z_axis) {
+			LOG_ERR("xyz-axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_X_MSB_RESULT;
+		end_address = TMAG5273_REG_Z_LSB_RESULT;
+		break;
+	case SENSOR_CHAN_DIE_TEMP:
+		if (!drv_cfg->temperature) {
+			LOG_ERR("temperature measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_T_MSB_RESULT;
+		end_address = TMAG5273_REG_T_LSB_RESULT;
+		break;
+	case SENSOR_CHAN_ROTATION:
+		if (drv_cfg->angle_magnitude_axis == TMAG5273_ANGLE_CALC_NONE) {
+			LOG_ERR("axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_ANGLE_MSB_RESULT;
+		end_address = TMAG5273_REG_ANGLE_LSB_RESULT;
+		break;
+	case TMAG5273_CHAN_MAGNITUDE:
+	case TMAG5273_CHAN_MAGNITUDE_MSB:
+		if (drv_cfg->angle_magnitude_axis == TMAG5273_ANGLE_CALC_NONE) {
+			LOG_ERR("axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = end_address = TMAG5273_REG_MAGNITUDE_RESULT;
+		break;
+	case TMAG5273_CHAN_ANGLE_MAGNITUDE:
+		if (drv_cfg->angle_magnitude_axis == TMAG5273_ANGLE_CALC_NONE) {
+			LOG_ERR("axis measurement deactivated");
+			return -ENOTSUP;
+		}
+		start_address = TMAG5273_REG_ANGLE_MSB_RESULT;
+		end_address = TMAG5273_REG_MAGNITUDE_RESULT;
+		break;
+	case SENSOR_CHAN_ALL:
+		start_address = TMAG5273_REG_RESULT_BEGIN;
+		end_address = TMAG5273_REG_RESULT_END;
+		break;
+	default:
+		LOG_ERR("unknown sensor channel %d", chan);
+		return -EINVAL;
+	}
+
+	uint32_t nb_bytes = end_address - start_address + 1;
+	uint8_t buf_offset = start_address - TMAG5273_REG_RESULT_BEGIN;
+
+	while (nb_bytes) {
+		const uint8_t block_size = COND_CODE_1(
+			CONFIG_CRC, ((drv_cfg->crc_enabled && (nb_bytes > 4)) ? 4 : nb_bytes),
+			(nb_bytes));
```

Comment:
that's a bit cryptic, how about moving the logic into a `tmag5273_get_block_size` and have normal `#if CONFIG_CRC` to branch the code?